src/node.rs

```diff
@@ -0,0 +1,171 @@
+// Provides user-level API for performing querie and interacting with the underlying service.
+#[derive(Debug, PartialEq)]
+pub struct Node {
+    pub node_id: Identifier,
+    pub table: KbucketTable,
+    pub store: HashMap<Vec<u8>, Vec<u8>>, // Same storage as Portal network to store samples
+}
+
+impl Node {
+    pub fn new(node_id: Identifier, table_record: TableRecord) -> Self {
+        Self {
+            node_id,
+            table: KbucketTable::new(node_id, table_record),
+            store: Default::default(),
+        }
+    }
+
+    // Protocol's RPCs:
+    // ---------------------------------------------------------------------------------------------------
+    /// "The most important procedure a Kademlia participant must perform is to locate
+    /// the k closest nodes to some given node ID"
+    ///     - Kademlia Paper
+    ///
+    /// Recieves an id request and returns node information on nodes within
+    /// *its closest bucket* (instead of k-closest nodes) to that id.
+    pub fn find_node(&mut self, id: &Identifier) -> HashMap<[u8; 32], TableRecord> {
+        self.table.get_bucket_for(id)
+    }
+
+    pub async fn ping(&mut self, local_socket: &UdpSocket, node_to_ping: &Identifier) -> usize {
+        let message_packet = b"Ping";
+
+        match self.table.get(node_to_ping) {
+            Some(remote_record) => {
+                let remote_socket =
+                    SocketAddrV4::new(remote_record.ip_address, remote_record.udp_port);
+                local_socket.connect(remote_socket).await;
+                local_socket.send(message_packet).await.unwrap()
+            }
+            _ => unreachable!("Node wasn't found to ping"),
+        }
+    }
+
+    // TODO:
+    // pub fn find_value() {}
+
+    /// "Most operations are implemented in terms of the lookup proceedure. To store a
+    /// <key,value> pair, a participant locates the k closes nodes to the key and sends them store RPCs".
+    ///
+    // TODO: Instructs a node to store a key, value pair for later retrieval.
+    // pub fn store(&mut self, key: Identifier, value: Vec<u8>) {}
+
+    // ---------------------------------------------------------------------------------------------------
+    pub async fn socket(&self) -> io::Result<UdpSocket> {
+        let table_record = self.table.record;
+        let socket_addr = SocketAddrV4::new(table_record.ip_address, table_record.udp_port);
+        let socket = UdpSocket::bind(socket_addr).await;
+        socket
+    }
+
+    // TODO: start_server() -> Result<(), Box<dyn std::error::Error>> {
+    pub async fn start_server(&self, mut buffer: [u8; 4], socket: UdpSocket) {
+        println!("Starting server");
+        loop {
+            let Ok((size, addr)) = socket.recv_from(&mut buffer).await else { todo!() };
+            let message = buffer.clone().to_vec();
+            self.process(message).await;
+        }
+    }
+
+    async fn process(&self, message: Vec<u8>) {
```

Comment:
cut your memory usage in half:

```suggestion
            self.process(&buffer).await;
        }
    }

    async fn process(&self, message: &[u8]) {
```

Reply:
Noob move...